packages/cache/src/Cache.ts

```diff
@@ -0,0 +1,204 @@
+import type { CacheKey, CacheOptions } from './types'
+import type { ExtractPartial } from './utility-types/ExtractPartial'
+import { hashCacheKey } from './utils'
+
+type Sync = (...args: unknown[]) => unknown
+
+export type AllStatusCached<TData, TCacheKey extends CacheKey = CacheKey> =
+  | {
+      status: 'idle'
+      promiseToSuspend: undefined
+      cacheKey: TCacheKey
+      hashedCacheKey: ReturnType<typeof hashCacheKey>
+      state: {
+        promise: undefined
+        data: undefined
+        error: undefined
+      }
+    }
+  | {
+      status: 'pending'
+      promiseToSuspend: Promise<void>
+      cacheKey: TCacheKey
+      hashedCacheKey: ReturnType<typeof hashCacheKey>
+      state: {
+        promise: Promise<TData>
+        data: undefined
+        error: undefined
+      }
+    }
+  | {
+      status: 'resolved'
+      promiseToSuspend: Promise<void>
+      cacheKey: TCacheKey
+      hashedCacheKey: ReturnType<typeof hashCacheKey>
+      state: {
+        promise: Promise<TData>
+        data: TData
+        error: undefined
+      }
+    }
+  | {
+      status: 'rejected'
+      promiseToSuspend: Promise<void>
+      cacheKey: TCacheKey
+      hashedCacheKey: ReturnType<typeof hashCacheKey>
+      state: {
+        promise: Promise<TData>
+        data: undefined
+        error: unknown
+      }
+    }
+
+export interface IdleCached<TData, TCacheKey extends CacheKey = CacheKey>
+  extends ExtractPartial<AllStatusCached<TData, TCacheKey>, { status: 'idle' }> {}
+export interface PendingCached<TData, TCacheKey extends CacheKey = CacheKey>
+  extends ExtractPartial<AllStatusCached<TData, TCacheKey>, { status: 'pending' }> {}
+export interface ResolvedCached<TData, TCacheKey extends CacheKey = CacheKey>
+  extends ExtractPartial<AllStatusCached<TData, TCacheKey>, { status: 'resolved' }> {}
+export interface RejectedCached<TData, TCacheKey extends CacheKey = CacheKey>
+  extends ExtractPartial<AllStatusCached<TData, TCacheKey>, { status: 'rejected' }> {}
+
+export type Cached<TData, TCacheKey extends CacheKey = CacheKey> =
+  | IdleCached<TData, TCacheKey>
+  | PendingCached<TData, TCacheKey>
+  | ResolvedCached<TData, TCacheKey>
+  | RejectedCached<TData, TCacheKey>
+
+/**
+ * @experimental This is experimental feature.
+ */
+export class Cache {
+  private cache = new Map<ReturnType<typeof hashCacheKey>, Cached<unknown>>()
+  private syncsMap = new Map<ReturnType<typeof hashCacheKey>, Array<Sync>>()
+
+  public reset = (options?: Pick<CacheOptions<unknown, CacheKey>, 'cacheKey'>) => {
+    if (typeof options?.cacheKey === 'undefined' || options.cacheKey.length === 0) {
+      this.cache.clear()
+      this.syncSubscribers()
+      return
+    }
+
+    const hashedKey = hashCacheKey(options.cacheKey)
```

Comment:
<img width="305" alt="image" src="https://github.com/user-attachments/assets/323647f7-06c0-4b7b-b706-38299522cba9">

Currently, there seems to be inconsistency in naming between 'hasedKey' and 'hasedCachKey'. A choice needs to be made.

Reply:
`hashedCacheKey` is better!

Reply:
I will make the changes. Thank you for choosing me!

Reply:
I modified it https://github.com/toss/suspensive/pull/1083/commits/3c501555f28544f73aeef798dab057b75029bcfb